base/range.jl

```diff
@@ -86,51 +77,139 @@ julia> range(1, 10, length=101)
 """
-range(start; length::Union{Integer,Nothing}=nothing, stop=nothing, step=nothing) =
+function range end
+
+range(start; stop=nothing, length::Union{Integer,Nothing}=nothing, step=nothing) =
     _range(start, step, stop, length)
 
-range(start, stop; length::Union{Integer,Nothing}=nothing, step=nothing) =
-    _range2(start, step, stop, length)
+function range(start, stop; length::Union{Integer,Nothing}=nothing, step=nothing)
+    # For code clarity, the user must pass step or length
+    # See https://github.com/JuliaLang/julia/pull/28708#issuecomment-420034562
+    if step === length === nothing
+        msg = """
+        Neither `step` nor `length` was provided. To fix this do one of the following:
+        * Pass one of them
+        * Use `$(start):$(stop)`
+        * Use `range($start, stop=$stop)`
+        """
+        throw(ArgumentError(msg))
+    end
+    _range(start, step, stop, length)
+end
 
-_range2(start, ::Nothing, stop, ::Nothing) =
-    throw(ArgumentError("At least one of `length` or `step` must be specified"))
+range(;start=nothing, stop=nothing, length::Union{Integer, Nothing}=nothing, step=nothing) =
+    _range(start, step, stop, length)
 
-_range2(start, step, stop, length) = _range(start, step, stop, length)
+_range(start::Nothing, step::Nothing, stop::Nothing, len::Nothing) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Nothing, stop::Nothing, len::Any    ) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Nothing, stop::Any    , len::Nothing) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Nothing, stop::Any    , len::Any    ) = range_stop_length(stop, len)
+_range(start::Nothing, step::Any    , stop::Nothing, len::Nothing) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Any    , stop::Nothing, len::Any    ) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Any    , stop::Any    , len::Nothing) = range_error(start, step, stop, len)
+_range(start::Nothing, step::Any    , stop::Any    , len::Any    ) = range_step_stop_length(step, stop, len)
+_range(start::Any    , step::Nothing, stop::Nothing, len::Nothing) = range_error(start, step, stop, len)
+_range(start::Any    , step::Nothing, stop::Nothing, len::Any    ) = range_start_length(start, len)
+_range(start::Any    , step::Nothing, stop::Any    , len::Nothing) = range_start_stop(start, stop)
+_range(start::Any    , step::Nothing, stop::Any    , len::Any    ) = range_start_stop_length(start, stop, len)
+_range(start::Any    , step::Any    , stop::Nothing, len::Nothing) = range_error(start, step, stop, len)
+_range(start::Any    , step::Any    , stop::Nothing, len::Any    ) = range_start_step_length(start, step, len)
+_range(start::Any    , step::Any    , stop::Any    , len::Nothing) = range_start_step_stop(start, step, stop)
+_range(start::Any    , step::Any    , stop::Any    , len::Any    ) = range_error(start, step, stop, len)
```

Comment:
Could we use ´@eval´ in a for loop here? This just seems a bit hard to maintain.

Reply:
I think it looks cool :stuck_out_tongue:

Reply:
It could use some comments. I think `@eval` would make it more complicated. There are four options and `4^2 == 16` lines. 16 lines with a non-trivial mapping seems pretty manageable to me. We need to make easier to read, not less.

How about something like this:

```julia
# range( )
_range(start::Nothing, step::Nothing, stop::Nothing, len::Nothing) = range_error(start, step, stop, len)
# range(; length )
_range(start::Nothing, step::Nothing, stop::Nothing, len::Any    ) = range_error(start, step, stop, len)
# range(; stop )
_range(start::Nothing, step::Nothing, stop::Any    , len::Nothing) = range_error(start, step, stop, len)
# range(; stop, length )
_range(start::Nothing, step::Nothing, stop::Any    , len::Any    ) = range_stop_length(stop, len)

Reply:
It looks very self-documenting to my eyes with a clear pattern.

Reply:
It is navigable to me, but I need to parse it to find the correct line. I'm happy with how it is though.

Reply:
I put some thought into how to write down this dispatch mechanic. I also considered using `@eval` but decided the above table is the most readable and easy to maintain.